Foxtrot/Driver/Rewriting/EmitAsyncClosure.cs

```diff
@@ -0,0 +1,969 @@
+
+            // Now, normal postconditions are written to the method body. 
+            // We need to add endOfNormalPrecondition block as a marker.
+            methodBodyBlock.Statements.Add(endOfNormalPrecondition);
+
+            //
+            // Exceptional postconditions
+            //
+
+            var exceptionalPostconditions = GetExceptionalEnsures(asyncPostconditions).ToList();
+            if (exceptionalPostconditions.Count > 0)
+            {
+                // For exceptional postconditions we need to generate CheckException method first
+                Method checkExceptionMethod = CreateCheckExceptionMethod();
+
+                EmitCheckExceptionBody(checkExceptionMethod, exceptionalPostconditions);
+
+                this.closureClass.Members.Add(checkExceptionMethod);
+
+                // Then, we're using the same trick as for normal postconditions:
+                // wrapping exceptional postconditions only when task.Status is TaskStatus.Faulted
+                Block checkExceptionBlock = new Block(new StatementList());
+
+                // Marker for EndOfNormalPrecondition
+                Block endOfExceptionPrecondition = new Block();
+
+                StatementList checkStatusIsException = CreateIfTaskResultIsEqualsTo(
+                    checkMethodTaskParameter, TaskStatus.Faulted, endOfExceptionPrecondition);
+
+                checkExceptionBlock.Statements.Add(new Block(checkStatusIsException));
+
+                // Now we need to emit actuall check for exceptional postconditions
+
+                // Emit: var ae = task.Exception;
+                
+                var aeLocal = new Local(aggregateExceptionType.Value);
+
+                checkExceptionBlock.Statements.Add(
+                    new AssignmentStatement(aeLocal,
+                        new MethodCall(
+                            new MemberBinding(checkMethodTaskParameter,
+                                GetTaskProperty(checkMethodTaskParameter, "get_Exception")),
+                            new ExpressionList())));
+
+                // Emit: CheckException(ae);
+
+                // Need to store method result somewhere, otherwise stack would be corrupted
+                var checkResultLocal = new Local(SystemTypes.Boolean);
+
+                checkExceptionBlock.Statements.Add(
+                    new AssignmentStatement(checkResultLocal,
+                            new MethodCall(new MemberBinding(null, 
+                                checkExceptionMethod),
+                            new ExpressionList(checkExceptionMethod.ThisParameter, aeLocal))));
+
+                checkExceptionBlock.Statements.Add(endOfExceptionPrecondition);
+                
+                methodBody.Add(checkExceptionBlock);
+            }
+
+            // Previously, following line was executed only without exceptional postconditions!
```

Comment:
:exclamation: If "previously" does not mean "earlier in this method", it should be removed. Maintain focus instead of the current behavior and the reason it is correct.